clients/postgres/src/main/scala/mqperf/postgres/PostgresMq.scala

```diff
@@ -0,0 +1,204 @@
+
+  override def init(config: Config): Unit = {
+    val table = config.mqConfig(TableConfigKey)
+    val client = databaseClient(config)
+
+    Await.result(
+      client
+        .sql(s"create table if not exists $table(id uuid primary key, content text not null, next_delivery timestamptz not null)")
+        .fetch()
+        .rowsUpdated()
+        .flatMap(_ =>
+          client
+            .sql(s"create index if not exists next_delivery_idx on $table(next_delivery)")
+            .fetch()
+            .rowsUpdated()
+        )
+        .toFuture
+        .asScala
+        .andThen {
+          case Success(_)  => logger.info(s"Table $table created.")
+          case Failure(ex) => logger.error(s"Table $table creating failure.", ex)
+        }
+        .map(_ => ()),
+      1.minute
+    )
+  }
+
+  override def cleanUp(config: Config): Unit = {
+    val table = config.mqConfig(TableConfigKey)
+
+    Await.result(
+      databaseClient(config)
+        .sql(s"drop table if exists $table")
+        .fetch()
+        .rowsUpdated()
+        .toFuture
+        .asScala
+        .andThen {
+          case Success(_)  => logger.info(s"Table $table dropped.")
+          case Failure(ex) => logger.error(s"Table $table dropping failure.", ex)
+        }
+        .map(_ => ()),
+      1.minute
+    )
+  }
+
+  override def createSender(config: Config): MqSender = new MqSender {
+
+    private val senderPool = connectionPool(config, config.mqConfig(SenderPoolSize).toInt)
+    private val client: DatabaseClient = pooledDatabaseClient(senderPool)
+
+    override def send(msgs: Seq[String]): Future[Unit] = {
+      val query = new StringBuilder("insert into jobs(id, content, next_delivery) values ")
+      query ++= s"('${UUID.randomUUID()}', '${msgs.head}', '${ZonedDateTime.now(clock)}')"
+      msgs.tail.foreach(m => query ++= s", ('${UUID.randomUUID()}', '$m', '${ZonedDateTime.now(clock)}')")
+
+      client
+        .sql(query.toString)
+        .fetch()
+        .rowsUpdated()
+        .toFuture
+        .asScala
+        .map(_ => ())
+    }
+
+    override def close(): Future[Unit] =
+      senderPool
+        .disposeLater()
+        .toFuture
+        .asScala
+        .map(_ => ())
+  }
+
+  override def createReceiver(config: Config): MqReceiver = new MqReceiver {
+
+    override type MsgId = UUID
+
+    private val receiverPool: ConnectionPool = connectionPool(config, config.mqConfig(ReceiverPoolSize).toInt)
+    private val client: DatabaseClient = pooledDatabaseClient(receiverPool)
+    private val txOperator: TransactionalOperator = TransactionalOperator.create(new R2dbcTransactionManager(receiverPool))
+
+    override def receive(maxMsgCount: Int): Future[Seq[(MsgId, String)]] = {
+      val now = ZonedDateTime.now(clock)
+      val nextDelivery = now.plusSeconds(100)
+
+      client
+        .sql("select id, content from jobs where next_delivery <= :now for update skip locked limit :limit")
```

Comment:
is this `.sql` result reusable, that is can we prepare this query once, and later only fill in the gaps?

Reply:
I think it is not now, but it must be possible. I will check it.